[FIXED] Unable to progress through workflow if "Only Allow Edit For" is Selected

javierwonghf opened the pull request:
This issue arose from PR https://github.com/frappe/frappe/pull/2477 where set_read_only actually removed all permissions for the entire document except read. This effectively gave the user no permissions to perform workflow actions and functions like duplicate.

This also fixes the problem where a user is unable to duplicate a document if it is part of a workflow (create permissions required).

rmehta commented on frappe/public/js/legacy/form.js:
```diff
@@ -922,6 +922,17 @@ _f.Frm.prototype.action_perm_type_map = {
+	// workflows if the user already have those permissions. This is to allow for users to 
+	// continue through the workflow states and to allow execution of functions like Duplicate.
+	if ((frappe.workflow.is_read_only(this.doctype, this.docname) && 
+		(frappe.perm.get_perm(this.doc.doctype)[0].write && perm_to_check == "write")) ||
```
set `perms = frappe.perm.get_perm(this.doc.doctype)[0]` so you call the function only once!

bohlian replied:
true true.. adjusted

javierwonghf committed:
[FIXED] Unable to progress through workflow if "Only Allow Edit For" is Selected
This also fixes the problem where a user is unable to duplicate a document if it is part of a workflow.

rmehta commented on frappe/public/js/legacy/form.js:
```diff
@@ -922,6 +922,16 @@ _f.Frm.prototype.action_perm_type_map = {
+	// continue through the workflow states and to allow execution of functions like Duplicate.
+	if (frappe.workflow.is_read_only(this.doctype, this.docname) && (perms["write"] ||
+		perms["create"] || perms["submit"])) {
+		var allowed_for_workflow = true;
```
where is this flag ultimately used?

bohlian replied:
When you configure a workflow with "Only Allow Edit for" set for a particular role. The entire form is set as read only. The thing is, set_read_only actually sets the read permission but strips all other permissions. 

So for a document which is subject to a workflow, a user with sufficient permissions would not be able to go to the next step of the workflow because the workflow action actually requires write permissions to store the workflow state. This is also the same for the "Duplicate" function which requires create permissions.

javierwonghf committed:
[UPDATED] Unable to progress through workflow if "Only Allow Edit For" is Selected.

bohlian commented:
Missed out the flag earlier. Updated to include it.
